Haskell/Haskell.sublime-syntax

```diff
@@ -1,407 +1,1388 @@
+    - match: then{{break}}
+      scope: keyword.control.conditional.then.haskell
+    - match: else{{break}}
+      scope: keyword.control.conditional.else.haskell
+    - match: let{{break}}
       scope: keyword.declaration.variable.haskell
-    - match: \breturn\b
-      scope: keyword.control.flow.return.haskell
-    - match: \b(?:default|otherwise)\b
-      scope: keyword.other.haskell
-    - match: \b(?:(if)|(then)|(else))\b
-      captures:
-        1: keyword.control.conditional.if.haskell
-        2: keyword.control.conditional.then.haskell
-        3: keyword.control.conditional.else.haskell
-
-  operator:
-    - match: (`)[a-zA-Z_'.\d]+(`)
-      # Haskell allows any ordinary function application (elem 4 [1..10])
-      # to be rewritten as an infix expression (4 `elem` [1..10])."
-      scope: keyword.operator.function.infix.haskell
-      captures:
-        1: punctuation.definition.function.begin.haskell
-        2: punctuation.definition.function.end.haskell
-    - match: '\binfix[lr]?\b'
+    - match: proc{{break}}
+      scope: keyword.declaration.function.haskell
+    # 4.4.2 Fixity Declarations
+    - match: infix[lr]?{{break}}
+      scope: keyword.declaration.fixity.haskell
+
+  operators:
+    - include: sequence-separators
+    - include: big-arrow-operators
+    - include: left-arrow-operators
+    - include: right-arrow-operators
+    - include: prefix-parens-operators
+    - include: infix-quoted-operators
+    # Match all not otherwise matched single quotes as promoition operator
+    # Note: Found in real world code but not in specs so far.
+    - match: \'
       scope: keyword.operator.haskell
-    - match: \b(\d+)(?:(\.)(\d+)([eE][-+]?\d+)?|([eE][-+]?\d+))\b
-      scope: meta.number.float.decimal.haskell
-      captures:
-        1: constant.numeric.value.haskell
-        2: punctuation.separator.decimal.haskell
-        3: constant.numeric.value.haskell
-        4: constant.numeric.value.exponent.haskell
-        5: constant.numeric.value.exponent.haskell
-    - match: '{{operator_infix}}'
+    - match: '{{operator_double_colon}}'
+      scope: keyword.operator.double-colon.haskell
+    - match: '{{operator_symbol}}'
       scope: keyword.operator.haskell
-    - include: operator_paren
 
-  operator_paren:
-    - match: \(({{operator_parens}})\)
-      scope: variable.function.infix.haskell
-      captures:
-        1: keyword.operator.haskell
+  big-arrow-operators:
+    - match: '{{operator_big_arrow}}'
+      scope: keyword.operator.big-arrow.haskell
 
-  preprocessor:
-    # In addition to Haskell's "native" syntax,
-    # GHC permits the C preprocessor to be run on a source file.
-    - match: ^\s*(#)\s*\w+
-      scope: meta.preprocessor.c
+  left-arrow-operators:
+    - match: '{{operator_left_arrow}}'
+      scope: keyword.operator.arrow.haskell
+
+  right-arrow-operators:
+    - match: '{{operator_right_arrow}}'
+      scope: keyword.operator.arrow.haskell
```

Comment:
I started to implement lambdas by specially scoping backslash followed by `variable.parameter` and also scoping the arrow in a special way. I rejected that afterall as that would have been the only expression with `variable.parameter` and finally decided to keep all kinds of operators colored the same way, too, as their meaning may differ by context as you already mension without a sane way to detect it due to language design.

I am not sure whether it would look off, if we'd scope the backslash `punctuation.definition` if parameters remain `variable.other`, but I don't have a very strong meaning about it.

Reply:
I certainly see value in semantically highlighting syntax operators like these arrows. Currently, this happens for the type annotations and declarations, so it would be a regression in that regard, but the others are not and can be considered suggestions or feature requests for a new ticket.

Same for `<-`.

Reply:
Semantic highlighting requires knowledge about context, but context boundaries are hard to maintain due to the nature of Haskells syntax. As python it relies on indentation to detect end of blocks. Even the old syntax did not correctly maintain type contexts in all situations, only for function signatures.

In order to not need to try to handle error prone context push-pops most semantic highlighting was removed with regards to operators and also type variables.

You may call it a regression, but it at least provides scope consistency for operators and variables all over the syntax in all situations even though less different scopes are used.

I tried to implement such things but stopped when I realized all kinds of implementation would end up in certain code styles to be needed for the syntax to work. And that's what I've learned from following VS Code's syntax.

Yes, semantic highlighting is nice, but not if it becomes the root cause of aweful heavy highlighting issues. That's not worth it.

Reply:
It's unfortunate if this is not feasible and while I do think it's a regression, it's definitely one that I can live with for a better overall experience. There might still be ways to target the most common type signatures and lambda definitions should definitely be possible, but I'm okay with the status quo.

Reply:
I split this into #2868.